Glsl.meta.slang khr shader subgroup

jkwak-work opened the pull request:
PR for a review purpose

ArielG-NV committed:
Partially Implement with tests, functions and built-in variables apart of GL_KHR_shader_subgroup; Partially resolves #3548

Partially Implement with tests, functions and built-in variables apart of GL_KHR_shader_subgroup; Partially resolves #3548

GL_KHR_shader_subgroup implemented based on https://github.com/KhronosGroup/GLSL/blob/main/extensions/khr/GL_KHR_shader_subgroup.txt

Implementation is broken down into seperate glsl extensions due to the ***large differences*** in implementation of each section, and functionality/testing.

GL_KHR_shader_subgroup_basic{
**Partially implemented**

Implementation:
    * All 9 built-in variables have been stubbed without proper value; implementation is still required for these system variables; related to #411.

    * Functions were reimplemented despite nearly mirrored HLSL functions due to:
        * hlsl.meta implementations targetting workgroups rather than a warp/wave/subgroup:
            * `__syncwarp` vs `__syncthreads`
            * `SubgroupMemory` vs `WorkgroupMemory`
            * etc.
        * hlsl.meta implementations target broader SPIR-V memory targets to block on:
            * ImageMemory|UniformMemory versus SPIR-V specifying barriers for ImageMemory and seperately an option for UniformMemory
        * `subgroupElect` for CUDA has a different implementation than `WaveIsFirstLane`, this is because spec states that `subgroupElect()` only returns the lowest active gl_SubgroupInvocationID; therefore we are supposed to fetch the current active mask even if some invocations are turned off by branches

Testing:
tests for the variable -- `tests/glsl/shader-subgroup-built-in-variables.slang`
    * these tests do not test functionality since not implemented yet

tests for the functions -- `tests/glsl/shader-subgroup-basic.slang`
    * concurrency is tested for using SubgroupMemory, UniformMemory through attempting to create a GPU side race condition with writing and reading memory
        * due to testing tools avaible there are no tests for ImageMemory
    * subgroupElect is tested to return invocation #0, the lowest invocation that will always run; wave size is 32, therefore #0 is always active and will always be the elected invocation.

}

GL_KHR_shader_subgroup_vote{
**Fully implemented**

Implementation:
    * 3/3 functions are using the hlsl.meta implementation

Testing:
`tests/glsl/shader-subgroup-vote.slang`
    * Testing each a positive (returns true) and negative (returns false) test case to ensure vote results are correct

}

GL_KHR_shader_subgroup_ballot{
**Partially implemented**

Implementation:
    There are 10/10 functions that are implemented:
    * 3 are using hlsl.meta implementation
    * 7 are using new implementations -- only support GLSL, SPIR-V, HLSL, CUDA
        * These implementations do not exist in hlsl.meta, so they were added
        * `subgroupInverseBallot` lacks an analog function to call; this feature was emulated:
            * in CUDA through knowing waves are 32bit and lanes are 0 indexed, this implys that `   (ballotResult >> YOUR_INVOCATION) & 1` checks if your invocation is active, for example, `(0b11001 >> 3) & 1` would mean that only invocation 5, 4, and 1 is active, 3 would mean `YOUR_INVOCATION` is the fourth invocation in the subgroup. `(0b11001>>3) & 1` would return true since your bit is toggled and evaluates to `0b11 & 0b1`
            * in HLSL through testing if the wave count is 32 or less (use the same logic as CUDA in this case); else find the index `YOUR_INVOCATION` corrisponds with where each vector has 32bits (32 waves); avoid division in the process. then run the same algorithm cuda employs.
            * `subgroupBallotBitExtract` is logically the same as `subgroupInverseBallot`
        * 5 implementations do not have a CUDA, HLSL, and CPP imlementation yet (subgroupBallotFindMSB, subgroupBallotFindLSB, subgroupBallotExclusiveBitCount, subgroupBallotInclusiveBitCount, subgroupBallotBitCount) due to being out of scope for the commit

Testing:
`tests/glsl/shader-subgroup-ballot.slang`
    * the function tests for an expected value of each ballot function; tests try inputting larger than 32 toggled bits as function parameters to ensure the implementation correctly identifies values up to a maximum of the subgroup invocation count as per extension specification (otherwise the functionality is fairly trivial to test)

}

GL_KHR_shader_subgroup_arithmetic{
**Partially implemented**

Implementation:
    * There are 21 functions to implement:
        * 14 functions are using the hlsl.meta implementation
        * 7 functions are new implementations -- only implemented for GLSL and SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required
            * CUDA, CPP, HLSL are out of scope for the commit

Testing:
`tests/glsl/shader-subgroup-arithmetic.slang`
    * all tests silently kill the shader; outputted GLSL was checked, could not see an issue
    * these tests only check basic functionality and correctness of all functions implemented; not an exaustive test [further continued in "Other notes of worthy" at end of commit]

}

GL_KHR_shader_subgroup_shuffle{
**Partially implemented**

Implementation:
    * There are 2 functions to implement:
        * 1 function is using the existing hlsl.meta implmentation
        * 1 function is using a new implmentation (subgroupShuffleXor) -- only implmented for GLSL & SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required

Testing:
`tests/glsl/shader-subgroup-shuffle.slang`
    * these tests only check basic functionality and correctness of all functions implemented; not an exaustive test [further continued in "Other notes of worthy" at end of commit]
    * tests fail with cpp due to `kIROp_WaveGetActiveMask` failing to be called

}

GL_KHR_shader_subgroup_shuffle_relative{
**Partially implemented**

Implementation:
    * There are 2 functions to implement:
        * all 2 functions are using a new implmentation -- only implmented for GLSL & SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required

Testing:
`tests/glsl/shader-subgroup-shuffle-relative.slang`
    * these tests only check basic functionality and correctness of all functions implemented; not an exaustive test [further continued in "Other notes of worthy" at end of commit]

}

GL_KHR_shader_subgroup_clustered{
**Partially implemented**

Implementation:
    * There are 7 functions to implement:
        * all 7 functions are using a new implmentation -- only implmented for GLSL & SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required

Testing:
`tests/glsl/shader-subgroup-shuffle-clustered.slang`
    * these tests only check basic functionality and correctness of all functions implemented; not an exaustive test [further continued in "Other notes of worthy" at end of commit]

}

GL_KHR_shader_subgroup_quad{
**Partially implemented**

Implementation:
    * There are 4 functions to implement:
        * all 4 functions are using hlsl.meta implmentations -- only implemented for GLSL & SPIR-V & HLSL

Testing:
`tests/glsl/shader-subgroup-shuffle-quad.slang`
    * these tests only check basic functionality and correctness of all functions implemented; not an exaustive test [further continued in "Other notes of worthy" at end of commit]

}

---------
Failing tests and why:

Note: due to system variables not being implemented largly for CUDA and CPP, these tests will fail (#3 and #4){
    tests/glsl/shader-subgroup-arithmetic.slang.3
    tests/glsl/shader-subgroup-arithmetic.slang.4
    tests/glsl/shader-subgroup-ballot.slang.4
    tests/glsl/shader-subgroup-basic.slang.3
    tests/glsl/shader-subgroup-basic.slang.4
    tests/glsl/shader-subgroup-quad.slang.3
    tests/glsl/shader-subgroup-quad.slang.4
    tests/glsl/shader-subgroup-vote.slang.3
    tests/glsl/shader-subgroup-vote.slang.4
}

Note: due to kIROp_WaveGetActiveMask not being loaded for cpp the following test will fail{
    tests/glsl/shader-subgroup-shuffle.slang.4
}

Note: due to a unknown silent error the following will fail [could not spot an error in the generated glsl and spir-v]{
    tests/glsl/shader-subgroup-arithmetic.slang.5 (vk)
    tests/glsl/shader-subgroup-arithmetic.slang.6 (vk)
}

Other notes of worthy:{

    * only a few types are checked currently in tests due to equality templates not allowing freely casting to int/uint, meaning to test types en-mass is not trivial and will most likley be completly replaced once templates can cast & check equality more freely.

    * did not implement vector types for any functions that may use them (mostly in reference to SPIR-V, since many may accept scalar or vector inputs); applicable to subgroup-shuffle, subgroup-shuffle-relative, subgroup-arithmetic, subgroup-shuffle, subgroup_clustered, subgroup_quad

    * did not implement checks for half floats

    * CUDA, CPP, HLSL implementations were largly out of scope and if not implemented, this is due to the implementation not being trivial

}

Random fixes encountered:{
    * hlsl.meta incorrectly sets `OpCapability` as `GroupNonUniformBallot` when the `OpCapability` should be `GroupNonUniformVote`; this is as per SPIR-V spec for all SPIR-V calls used in `GL_KHR_shader_subgroup_vote`: https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpGroupNonUniformAll
}

jkwak-work commented on source/slang/glsl.meta.slang:
````diff
@@ -417,3 +417,1399 @@ ${{{{
+```
+*/
+[ForceInline] bool __isMinimumWarpActiveInBranchCUDA() {
+    __intrinsic_asm "( (__activemask() & __activemask()*-1) == _getLaneId())";
````
Can you add one more parenthesis?
(__activemask() & (__activemask() * -1))

ArielG-NV replied:
👍

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -417,3 +417,1399 @@ ${{{{
+int main(){
+    printf("%b\n", isMin(v1));
+    printf("%b\n", isMin(v2));
+}
```
Let's remove the testing code above.

ArielG-NV replied:
👍

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -417,3 +417,1399 @@ ${{{{
+__spirv_version(1.3)
+[ForceInline] T __subgroupBroadcastSPIRV(T value, uint id)
+{
+    return (spirv_asm {
```
Let's not have an extra parenthesis before spirv_asm.
return spirv_asm {

ArielG-NV replied:
👍

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -417,3 +417,1399 @@ ${{{{
+public void subgroupMemoryBarrierBuffer();
+public void subgroupMemoryBarrierImage();
+public void subgroupMemoryBarrierShared();
+public bool subgroupElect();
```
I don't think we need forward declarations for those functions.
You may want to put "public" keywords on each functions tho.
[ForceInline] public void subgroupBarrier()

ArielG-NV replied:
I tested in a various method of ways; the syntax does not parse correctly for glsl, public on forward declarations is the only method of definition that worked for me 
![image](https://github.com/jkwak-work/slang/assets/159081215/65b26170-1459-42d0-935e-047ee9a76dcc)

jkwak-work replied:
It appeas that the error message for it is
```
glsl.meta.slang(2195): error 36103: public symbol 'subgroupBarrier' is missing capability requirement declaration.
[ForceInline] public void subgroupBarrier()
```

jkwak-work replied:
Check the Slack conversation about this issue,
https://nvidia.slack.com/archives/CFFF96M6Z/p1707210299287209

ArielG-NV replied:
👍

ArielG-NV committed:
Partially Implement with tests, functions and built-in variables apart of GL_KHR_shader_subgroup; Partially resolves #3548

Partially Implement with tests, functions and built-in variables apart of GL_KHR_shader_subgroup; Partially resolves #3548

GL_KHR_shader_subgroup implemented based on https://github.com/KhronosGroup/GLSL/blob/main/extensions/khr/GL_KHR_shader_subgroup.txt

GL_KHR_shader_subgroup_* & GLSL ref:
    * https://github.com/KhronosGroup/GLSL/blob/main/extensions/khr/GL_KHR_shader_subgroup.txt
    * https://www.khronos.org/blog/vulkan-subgroup-tutorial
    * https://www.khronos.org/assets/uploads/developers/library/2018-vulkan-devday/06-subgroups.pdf

HLSL ref:
    * https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-intrinsic-functions
    * https://github.com/Microsoft/DirectXShaderCompiler/wiki/Wave-Intrinsics

CUDA ref:
    * https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html

SPIR-V ref:
    * https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_memory_semantics_id

Implementation is broken down into seperate glsl extensions due to the ***large differences*** in implementation of each section, and functionality/testing.

GL_KHR_shader_subgroup_basic{
**Partially implemented**

Implementation:
    * All 9 built-in variables have been stubbed without proper value; implementation is still required for these system variables; related to #411.

    * Functions were reimplemented despite nearly mirrored HLSL functions due to:
        * hlsl.meta implementations targetting workgroups rather than a warp/wave/subgroup:
            * `__syncwarp` vs `__syncthreads`
            * `SubgroupMemory` vs `WorkgroupMemory`
            * etc.
        * hlsl.meta implementations target broader SPIR-V memory targets to block on:
            * ImageMemory|UniformMemory versus SPIR-V specifying barriers for ImageMemory and seperately an option for UniformMemory
        * `subgroupElect` for CUDA has a different implementation than `WaveIsFirstLane`, this is because spec states that `subgroupElect()` only returns the lowest active gl_SubgroupInvocationID; therefore we are supposed to fetch the current active mask even if some invocations are turned off by branches

Testing:
tests for the variable -- `tests/glsl/shader-subgroup-built-in-variables.slang`
    * these tests do not test functionality since not implemented yet

tests for the functions -- `tests/glsl/shader-subgroup-basic.slang`
    * concurrency is tested for using SubgroupMemory, UniformMemory through attempting to create a GPU side race condition with writing and reading memory
        * due to testing tools avaible there are no tests for ImageMemory
    * subgroupElect is tested to return invocation #0, the lowest invocation that will always run; wave size is 32, therefore #0 is always active and will always be the elected invocation.

}

GL_KHR_shader_subgroup_vote{
**Fully implemented**

Implementation:
    * 3/3 functions are using the hlsl.meta implementation

Testing:

`tests/glsl/shader-subgroup-vote.slang`
    * Testing each a positive (returns true) and negative (returns false) test case to ensure vote results are correct

}

GL_KHR_shader_subgroup_ballot{
**Partially implemented**

Implementation:
    There are 10/10 functions that are implemented:
    * 3 are using hlsl.meta implementation
    * 7 are using new implementations -- only support GLSL, SPIR-V, HLSL, CUDA
        * These implementations do not exist in hlsl.meta, so they were added
        * `subgroupInverseBallot` lacks an analog function to call; this feature was emulated:
            * in CUDA through knowing waves are 32bit and lanes are 0 indexed, this implys that `   (ballotResult >> YOUR_INVOCATION) & 1` checks if your invocation is active, for example, `(0b11001 >> 3) & 1` would mean that only invocation 5, 4, and 1 is active, 3 would mean `YOUR_INVOCATION` is the fourth invocation in the subgroup. `(0b11001>>3) & 1` would return true since your bit is toggled and evaluates to `0b11 & 0b1`
            * in HLSL through testing if the wave count is 32 or less (use the same logic as CUDA in this case); else find the index `YOUR_INVOCATION` corrisponds with where each vector has 32bits (32 waves); avoid division in the process. then run the same algorithm cuda employs.
            * `subgroupBallotBitExtract` is logically the same as `subgroupInverseBallot`
        * 5 implementations do not have a CUDA, HLSL, and CPP imlementation yet (subgroupBallotFindMSB, subgroupBallotFindLSB, subgroupBallotExclusiveBitCount, subgroupBallotInclusiveBitCount, subgroupBallotBitCount) due to being out of scope for the commit

Testing:
`tests/glsl/shader-subgroup-ballot.slang`
    * the function tests for an expected value of each ballot function; tests try inputting larger than 32 toggled bits as function parameters to ensure the implementation correctly identifies values up to a maximum of the subgroup invocation count as per extension specification (otherwise the functionality is fairly trivial to test)

}

GL_KHR_shader_subgroup_arithmetic{
**Partially implemented**

Implementation:
    * There are 21 functions to implement:
        * 14 functions are using the hlsl.meta implementation
        * 7 functions are new implementations -- only implemented for GLSL and SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required
            * CUDA, CPP, HLSL are out of scope for the commit

Testing:
`tests/glsl/shader-subgroup-arithmetic.slang`
    * all tests silently kill the shader; outputted GLSL was checked, could not see an issue
    * these tests only check basic functionality and correctness of all functions implemented; [further continued in "Other notes of worthy" at end of commit]

}

GL_KHR_shader_subgroup_shuffle{
**Partially implemented**

Implementation:
    * There are 2 functions to implement:
        * 1 function is using the existing hlsl.meta implmentation
        * 1 function is using a new implmentation (subgroupShuffleXor) -- only implmented for GLSL & SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required

Testing:
`tests/glsl/shader-subgroup-shuffle.slang`
    * these tests only check basic functionality and correctness of all functions implemented; [further continued in "Other notes of worthy" at end of commit]
    * tests fail with cpp due to `kIROp_WaveGetActiveMask` failing to be called

}

GL_KHR_shader_subgroup_shuffle_relative{
**Partially implemented**

Implementation:
    * There are 2 functions to implement:
        * all 2 functions are using a new implmentation -- only implmented for GLSL & SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required

Testing:
`tests/glsl/shader-subgroup-shuffle-relative.slang`
    * these tests only check basic functionality and correctness of all functions implemented; [further continued in "Other notes of worthy" at end of commit]

}

GL_KHR_shader_subgroup_clustered{
**Partially implemented**

Implementation:
    * There are 7 functions to implement:
        * all 7 functions are using a new implmentation -- only implmented for GLSL & SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required

Testing:
`tests/glsl/shader-subgroup-shuffle-clustered.slang`
    * these tests only check basic functionality and correctness of all functions implemented; [further continued in "Other notes of worthy" at end of commit]

}

GL_KHR_shader_subgroup_quad{
**Partially implemented**

Implementation:
    * There are 4 functions to implement:
        * all 4 functions are using hlsl.meta implmentations -- only implemented for GLSL & SPIR-V & HLSL

Testing:
`tests/glsl/shader-subgroup-shuffle-quad.slang`
    * these tests only check basic functionality and correctness of all functions implemented; [further continued in "Other notes of worthy" at end of commit]

}

---------
Failing tests and why:

Note: test numbers are assuming none of the existing tests are toggled off

Note: due to system variables not being implemented largly for CUDA and CPP, these tests will fail (#3 and #4){
    tests/glsl/shader-subgroup-arithmetic.slang.3
    tests/glsl/shader-subgroup-arithmetic.slang.4
    tests/glsl/shader-subgroup-ballot.slang.4
    tests/glsl/shader-subgroup-basic.slang.3
    tests/glsl/shader-subgroup-basic.slang.4
    tests/glsl/shader-subgroup-quad.slang.3
    tests/glsl/shader-subgroup-quad.slang.4
    tests/glsl/shader-subgroup-vote.slang.3
    tests/glsl/shader-subgroup-vote.slang.4
}

Note: due to kIROp_WaveGetActiveMask not being loaded for cpp the following test will fail{
    tests/glsl/shader-subgroup-shuffle.slang.4
    tests/glsl/shader-subgroup-shuffle-relative.slang.4
    tests/glsl/shader-subgroup-basic.slang.4
}

Other notes of worthy:{

    * added preamble function and macros for implementing subgroup functionality (and tests) to make it possible to iterate on the functionality with reasonable effort in the future

    * CUDA, CPP, HLSL implementations were largly out of scope and not implemented, this is due to the implementation being non trivial for many functions

    * doubles cause a silent crash on most subgroup functions tested (silent shader hang)

    * __requireGLSLExtension does not work as intended inside glsl.meta; as a result half, int16, int64 int8, all are ommited from testing

}

Random fixes encountered:{
    * hlsl.meta incorrectly sets `OpCapability` as `GroupNonUniformBallot` when the `OpCapability` should be `GroupNonUniformVote`; this is as per SPIR-V spec for all SPIR-V calls used in `GL_KHR_shader_subgroup_vote`: https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpGroupNonUniformAll
    * hlsl.meta incorrectly uses for WaveMaskPrefixBitOr (SPIR-V) OpGroupNonUniformBitwiseAnd intead of OpGroupNonUniformBitwiseOr; this was fixed
}

ArielG-NV committed:
added vector types and tests;

Partially Implement with tests, functions and built-in variables apart of GL_KHR_shader_subgroup; Partially resolves #3548

GL_KHR_shader_subgroup implemented based on https://github.com/KhronosGroup/GLSL/blob/main/extensions/khr/GL_KHR_shader_subgroup.txt

GL_KHR_shader_subgroup_* & GLSL ref:
    * https://github.com/KhronosGroup/GLSL/blob/main/extensions/khr/GL_KHR_shader_subgroup.txt
    * https://www.khronos.org/blog/vulkan-subgroup-tutorial
    * https://www.khronos.org/assets/uploads/developers/library/2018-vulkan-devday/06-subgroups.pdf

HLSL ref:
    * https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-intrinsic-functions
    * https://github.com/Microsoft/DirectXShaderCompiler/wiki/Wave-Intrinsics

CUDA ref:
    * https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html

SPIR-V ref:
    * https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_memory_semantics_id

Implementation is broken down into seperate glsl extensions due to the ***large differences*** in implementation of each section, and functionality/testing.

GL_KHR_shader_subgroup_basic{
**Partially implemented**

Implementation:
    * All 9 built-in variables have been stubbed without proper value; implementation is still required for these system variables; related to #411.

    * Functions were reimplemented despite nearly mirrored HLSL functions due to:
        * hlsl.meta implementations targetting workgroups rather than a warp/wave/subgroup:
            * `__syncwarp` vs `__syncthreads`
            * `SubgroupMemory` vs `WorkgroupMemory`
            * etc.
        * hlsl.meta implementations target broader SPIR-V memory targets to block on:
            * ImageMemory|UniformMemory versus SPIR-V specifying barriers for ImageMemory and seperately an option for UniformMemory
        * `subgroupElect` for CUDA has a different implementation than `WaveIsFirstLane`, this is because spec states that `subgroupElect()` only returns the lowest active gl_SubgroupInvocationID; therefore we are supposed to fetch the current active mask even if some invocations are turned off by branches

Testing:
tests for the variable -- `tests/glsl/shader-subgroup-built-in-variables.slang`
    * these tests do not test functionality since not implemented yet

tests for the functions -- `tests/glsl/shader-subgroup-basic.slang`
    * concurrency is tested for using SubgroupMemory, UniformMemory through attempting to create a GPU side race condition with writing and reading memory
        * due to testing tools avaible there are no tests for ImageMemory
    * subgroupElect is tested to return invocation #0, the lowest invocation that will always run; wave size is 32, therefore #0 is always active and will always be the elected invocation.

}

GL_KHR_shader_subgroup_vote{
**Fully implemented**

Implementation:
    * 3/3 functions are using the hlsl.meta implementation

Testing:

`tests/glsl/shader-subgroup-vote.slang`
    * Testing each a positive (returns true) and negative (returns false) test case to ensure vote results are correct

}

GL_KHR_shader_subgroup_ballot{
**Partially implemented**

Implementation:
    There are 10/10 functions that are implemented:
    * 3 are using hlsl.meta implementation
    * 7 are using new implementations -- only support GLSL, SPIR-V, HLSL, CUDA
        * These implementations do not exist in hlsl.meta, so they were added
        * `subgroupInverseBallot` lacks an analog function to call; this feature was emulated:
            * in CUDA through knowing waves are 32bit and lanes are 0 indexed, this implys that `   (ballotResult >> YOUR_INVOCATION) & 1` checks if your invocation is active, for example, `(0b11001 >> 3) & 1` would mean that only invocation 5, 4, and 1 is active, 3 would mean `YOUR_INVOCATION` is the fourth invocation in the subgroup. `(0b11001>>3) & 1` would return true since your bit is toggled and evaluates to `0b11 & 0b1`
            * in HLSL through testing if the wave count is 32 or less (use the same logic as CUDA in this case); else find the index `YOUR_INVOCATION` corrisponds with where each vector has 32bits (32 waves); avoid division in the process. then run the same algorithm cuda employs.
            * `subgroupBallotBitExtract` is logically the same as `subgroupInverseBallot`
        * 5 implementations do not have a CUDA, HLSL, and CPP imlementation yet (subgroupBallotFindMSB, subgroupBallotFindLSB, subgroupBallotExclusiveBitCount, subgroupBallotInclusiveBitCount, subgroupBallotBitCount) due to being out of scope for the commit

Testing:
`tests/glsl/shader-subgroup-ballot.slang`
    * the function tests for an expected value of each ballot function; tests try inputting larger than 32 toggled bits as function parameters to ensure the implementation correctly identifies values up to a maximum of the subgroup invocation count as per extension specification (otherwise the functionality is fairly trivial to test)

}

GL_KHR_shader_subgroup_arithmetic{
**Partially implemented**

Implementation:
    * There are 21 functions to implement:
        * 14 functions are using the hlsl.meta implementation
        * 7 functions are new implementations -- only implemented for GLSL and SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required
            * CUDA, CPP, HLSL are out of scope for the commit

Testing:
`tests/glsl/shader-subgroup-arithmetic.slang`
    * all tests silently kill the shader; outputted GLSL was checked, could not see an issue
    * these tests only check basic functionality and correctness of all functions implemented; [further continued in "Other notes of worthy" at end of commit]

}

GL_KHR_shader_subgroup_shuffle{
**Partially implemented**

Implementation:
    * There are 2 functions to implement:
        * 1 function is using the existing hlsl.meta implmentation
        * 1 function is using a new implmentation (subgroupShuffleXor) -- only implmented for GLSL & SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required

Testing:
`tests/glsl/shader-subgroup-shuffle.slang`
    * these tests only check basic functionality and correctness of all functions implemented; [further continued in "Other notes of worthy" at end of commit]
    * tests fail with cpp due to `kIROp_WaveGetActiveMask` failing to be called

}

GL_KHR_shader_subgroup_shuffle_relative{
**Partially implemented**

Implementation:
    * There are 2 functions to implement:
        * all 2 functions are using a new implmentation -- only implmented for GLSL & SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required

Testing:
`tests/glsl/shader-subgroup-shuffle-relative.slang`
    * these tests only check basic functionality and correctness of all functions implemented; [further continued in "Other notes of worthy" at end of commit]

}

GL_KHR_shader_subgroup_clustered{
**Partially implemented**

Implementation:
    * There are 7 functions to implement:
        * all 7 functions are using a new implmentation -- only implmented for GLSL & SPIR-V
            * GLSL & SPIR-V both use their related functions, no emulation required

Testing:
`tests/glsl/shader-subgroup-shuffle-clustered.slang`
    * these tests only check basic functionality and correctness of all functions implemented; [further continued in "Other notes of worthy" at end of commit]

}

GL_KHR_shader_subgroup_quad{
**Partially implemented**

Implementation:
    * There are 4 functions to implement:
        * all 4 functions are using hlsl.meta implmentations -- only implemented for GLSL & SPIR-V & HLSL

Testing:
`tests/glsl/shader-subgroup-shuffle-quad.slang`
    * these tests only check basic functionality and correctness of all functions implemented; [further continued in "Other notes of worthy" at end of commit]

}

---------
Failing tests and why:

Note: test numbers are assuming none of the existing tests are toggled off

Note: due to system variables not being implemented largly for CUDA and CPP, these tests will fail (#3 and #4){
    tests/glsl/shader-subgroup-arithmetic.slang.3
    tests/glsl/shader-subgroup-arithmetic.slang.4
    tests/glsl/shader-subgroup-ballot.slang.4
    tests/glsl/shader-subgroup-basic.slang.3
    tests/glsl/shader-subgroup-basic.slang.4
    tests/glsl/shader-subgroup-quad.slang.3
    tests/glsl/shader-subgroup-quad.slang.4
    tests/glsl/shader-subgroup-vote.slang.3
    tests/glsl/shader-subgroup-vote.slang.4
}

Note: due to kIROp_WaveGetActiveMask not being loaded for cpp the following test will fail{
    tests/glsl/shader-subgroup-shuffle.slang.4
    tests/glsl/shader-subgroup-shuffle-relative.slang.4
    tests/glsl/shader-subgroup-basic.slang.4
}

Note: due to a unknown silent error the following will fail [could not spot an error in the generated glsl and spir-v]{
    tests/glsl/shader-subgroup-arithmetic.slang.5 (vk)
    tests/glsl/shader-subgroup-arithmetic.slang.6 (vk)
}

Other notes of worthy:{

    * only a few types are checked currently in arithmetic test; this is due to the test silently failing, meaning I can't actually test anything implemented

    * did not implement checks for half floats

    * CUDA, CPP, HLSL implementations were largly out of scope and not implemented, this is due to the implementation being non trivial for many functions

}

Random fixes encountered:{
    * hlsl.meta incorrectly sets `OpCapability` as `GroupNonUniformBallot` when the `OpCapability` should be `GroupNonUniformVote`; this is as per SPIR-V spec for all SPIR-V calls used in `GL_KHR_shader_subgroup_vote`: https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpGroupNonUniformAll
}

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -417,3 +417,1399 @@ ${{{{
+    return WaveMaskAllEqual(WaveGetActiveMask(), value);
+}
+
+// ~GL_KHR_shader_subgroup_vote
```
I don't think the project uses this notation.
Let's remove these.

ArielG-NV replied:
👍

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -417,3 +417,1399 @@ ${{{{
+
+__glsl_extension(GL_KHR_shader_subgroup_vote)
+__spirv_version(1.3)
+[ForceInline] bool subgroupAllEqual(bool value)
```
The spec says that the signature for this function is for multiple types not just for bool.
        bool subgroupAllEqual(genType value);
        bool subgroupAllEqual(genIType value);
        bool subgroupAllEqual(genUType value);
        bool subgroupAllEqual(genBType value);
        bool subgroupAllEqual(genDType value);

https://github.com/KhronosGroup/GLSL/blob/886df1b2b544ade9076fcf4675f2ecbf2963dd86/extensions/khr/GL_KHR_shader_subgroup.txt#L763

ArielG-NV replied:
👍

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -417,3 +417,1399 @@ ${{{{
+__generic<T : __BuiltinArithmeticType>
+__glsl_extension(GL_KHR_shader_subgroup_arithmetic)
+__spirv_version(1.3)
+[ForceInline] T subgroupAdd(T value)
```
Can you list functions in the order that they appear in the spec?
It seems like subgroupShuffle should be here.

ArielG-NV replied:
changed function definition order to match https://github.com/KhronosGroup/GLSL/blob/886df1b2b544ade9076fcf4675f2ecbf2963dd86/extensions/khr/GL_KHR_shader_subgroup.txt#L763

jkwak-work commented on tests/glsl/shader-subgroup-shuffle.slang:
```diff
@@ -0,0 +1,45 @@
+    // TODO: make more robust of a test than simple call verification
+
+
+    outputBuffer.data[0] = subgroupShuffle(1, 1) == 1;
```
According to the spec, subgroupShuffle has following signature,
        genType  subgroupShuffle(genType value,  uint id);
        genIType subgroupShuffle(genIType value, uint id);
        genUType subgroupShuffle(genUType value, uint id);
        genBType subgroupShuffle(genBType value, uint id);
        genDType subgroupShuffle(genDType value, uint id);
Note that the OpenGL spec doc explains what genXType means, https://registry.khronos.org/OpenGL/specs/gl/GLSLangSpec.4.60.pdf
```
Where the input arguments (and corresponding output) can be int, ivec2, ivec3, or ivec4, genIType is used as the argument.
```
You will have to call subgroupShuffle with all of the possible types and make sure they exist.

ArielG-NV replied:
will be in the next commit; note, as per commit vectors are not implemented, doubles crash

ArielG-NV replied:
(added to shuffle and shuffle relative)

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -417,3 +417,1399 @@ ${{{{
+            return spirv_asm
+            {
+                OpCapability GroupNonUniformArithmetic; OpCapability GroupNonUniformClustered;
+                // TODO: use the correct integer width
```
I don't think you can use a comment notation inside of spirv_asm.
Or at least we don't do it as far as I know.

ArielG-NV replied:
this was code based off of the hlsl.meta, it works fine in tests

jkwak-work reviewed (commented):
I have some comments on the code

jkwak-work commented on tests/glsl/shader-subgroup-arithmetic.slang:
```diff
@@ -0,0 +1,158 @@
+//    outputBuffer.data[index + 19] = subgroupInclusiveOr<M>(test) == 1;
+//    outputBuffer.data[index + 20] = subgroupInclusiveXor<M>(test) == 1;
+    
+//}
```
Can you describe why this code is commented out?
Please check out how I disabled certain parts of the code with "#if defined".
https://github.com/shader-slang/slang/blob/master/tests/glsl-intrinsic/intrinsic-basic.slang

ArielG-NV replied:
👍

jkwak-work commented on tests/glsl/shader-subgroup-arithmetic.slang:
```diff
@@ -0,0 +1,158 @@
+    // testBasedOnType(1, (int)1, (uint)1);
+    // testBasedOnType(2, (uint)1, (int)1);
+    // testBasedOnType(3, (double)1, (uint)1);
+    // testBasedOnType(4, (half)1, (int)1);
```
Based on the comment at line 77, these code block is commented out because of the error with CUDA and CPP.
In this case, you can give -DTARGET_CUDA and -DTARGET_CPP to the "//TEST" lines at the top and wrap them like following,
```
//TEST:SIMPLE(filecheck=CHECK_CUDA): -allow-glsl -stage compute -entry computeMain -target cuda -DTARGET_CUDA
//TEST:SIMPLE(filecheck=CHECK_CPP): -allow-glsl -stage compute -entry computeMain -target cpp -DTARGET_CPP
```
and
```
#if !defined(TARGET_CUDA) && !defined(TARGET_CPP)
```

ArielG-NV replied:
the errors are a problem with all versions of the code

jkwak-work reviewed (commented):
More comments

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -461,11 +460,6 @@ public uint gl_SubgroupLeMask;
-
-/// GL_KHR_shader_subgroup Functions
-
-// GL_KHR_shader_subgroup_basic
```
I thought this was a good comment.
I meant to remove /// ~XXX that has the tide(~) character.

ArielG-NV replied:
👍

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -403,7 +403,7 @@ public bool operator!=<let N : int>(vector<$(typeName), N> left, vector<$(typeNa
     return any(notEqual(left, right));
 }
-${{{{
+${{{{ 
```
This doesn't need a change.
A tailing white character is not good anyway.

ArielG-NV replied:
👍

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -417,3 +417,1092 @@ ${{{{
+//
+// CUDA ref: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html
+//
+// SPIR-V ref: https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_memory_semantics_id
```
I think these comments should go to the commit message not in the source code.

ArielG-NV replied:
👍

jkwak-work commented on tests/glsl/shader-subgroup-arithmetic.slang:
```diff
@@ -0,0 +1,296 @@
+// TEST(compute, vulkan):COMPARE_COMPUTE(filecheck-buffer=BUF):-vk -compute -entry computeMain -allow-glsl -emit-spirv-directly
+#version 430
+
+// #define TEST_when_equallity_and_uint_cast_on_templates_work
```
Does this need to be disabled for all targets??? Even for glsl?
You can selectively enable it for glsl if you should.
```
#if defined(TARGET_GLSL)
#define TEST_when_equallity_and_uint_cast_on_templates_work
#endif
```

jkwak-work replied:
And what is "uint cast on templates"?

ArielG-NV replied:
for templates, doing something like call   `uint v = myCall<T>(val) == 1; ` will fail since if the result is type T, equality was not implemented for templates (this won't work)

ArielG-NV replied:
> Does this need to be disabled for all targets??? Even for glsl? You can selectively enable it for glsl if you should.
> 
> ```
> #if defined(TARGET_GLSL)
> #define TEST_when_equallity_and_uint_cast_on_templates_work
> #endif
> ```

this problem effects all

jkwak-work commented on tests/glsl/shader-subgroup-arithmetic.slang:
```diff
@@ -0,0 +1,296 @@
+    
+    outputBuffer.data[index + 4] = subgroupAnd<M>(test2) == 0;
+    outputBuffer.data[index + 5] = subgroupOr<M>(test2) == 63;
+    outputBuffer.data[index + 6] = subgroupXor<M>(test2) == 0;
```
These are not a correctly syntax for GLSL.
You should call them without the template argument N.
```
subgroupAdd(test) == 32;
```

ArielG-NV replied:
👍 the tests don't work regardless, but I did remove the template attribute specifier (for the next commit)

jkwak-work commented on tests/glsl/shader-subgroup-arithmetic.slang:
```diff
@@ -0,0 +1,296 @@
+     testBasedOnType(1, (int)1, (uint)1);
+     testBasedOnType(2, (uint)1, (int)1);
+     testBasedOnType(3, (double)1, (uint)1);
+     testBasedOnType(4, (half)1, (int)1);
```
This looks confusing.
Let's split the function into two functions:
```
arithmaticTest(0, float(1));
arithmaticTest(1, double(1));
arithmaticTest(2, half(1));
arithmaticTest(3, int(1));
arithmaticTest(4, uint(1));
integerTest(5,int(1));
integerTest(6, uint(1));
```

ArielG-NV replied:
I don't know if I should even keep the tests since they don't and won't work until templates have new functionality (even if I separate the tests, to check if they even compile/work would be a task for the future)

jkwak-work commented on tests/glsl/shader-subgroup-arithmetic.slang:
```diff
@@ -0,0 +1,296 @@
+// not testing glsl or spirv due to crashing issue from some odd code emitting which looks right but fails
+// TEST:SIMPLE(filecheck=CHECK_GLSL):  -allow-glsl -stage compute -entry computeMain -target glsl
```
This source code is in GLSL syntax.
It doesn't make sense that you cannot use GLSL as target for this test.
GLSL -> GLSL should work.

ArielG-NV replied:
the emitted code is read by slang incorrectly I think: attached is a picture of normal glsl which would compile on any normal glsl compiler, but on slang it does not
![image](https://github.com/jkwak-work/slang/assets/159081215/32d104e3-a35d-4c76-9f0f-3880899f39dd)

jkwak-work commented on tests/glsl/shader-subgroup-ballot.slang:
```diff
@@ -0,0 +1,62 @@
+// TEST:SIMPLE(filecheck=CHECK_CPP):  -allow-glsl -stage compute -entry computeMain -target cpp
+
+//TEST(compute, vulkan):COMPARE_COMPUTE(filecheck-buffer=BUF):-vk -compute -entry computeMain -allow-glsl
+//TEST(compute, vulkan):COMPARE_COMPUTE(filecheck-buffer=BUF):-vk -compute -entry computeMain -allow-glsl -emit-spirv-directly
```
These two tests are failing for me.

ArielG-NV replied:
I don't know why, it works for me

jkwak-work replied:
I am getting an error message as following and I guess there is a problem at the line 59
```
error: slang-test: tests/glsl/shader-subgroup-ballot.slang:59:15: error: BUF-NEXT: is not on the line after the previous match
```

jkwak-work commented on tests/glsl/shader-subgroup-basic.slang:
```diff
@@ -0,0 +1,74 @@
+// TEST:SIMPLE(filecheck=CHECK_CPP):  -allow-glsl -stage compute -entry computeMain -target cpp
+
+//TEST(compute, vulkan):COMPARE_COMPUTE(filecheck-buffer=BUF):-vk -compute -entry computeMain -allow-glsl
+//TEST(compute, vulkan):COMPARE_COMPUTE(filecheck-buffer=BUF):-vk -compute -entry computeMain -allow-glsl -emit-spirv-directly
```
This test is failing for me.

ArielG-NV replied:
I don't know why, it works for me

jkwak-work replied:
The error message is following.
```
slang-test: tests/glsl/shader-subgroup-basic.slang:69:10: error: BUF: expected string not found in input
```

jkwak-work commented on tests/glsl/shader-subgroup-builtin-variables.slang:
```diff
@@ -0,0 +1,61 @@
+    uint data[];
+} outputBuffer;
+
+layout(local_size_x = 32) in;
```
You don't seem to need these in this test case.

ArielG-NV replied:
I don't but the tests do need to be changed and to actually test data later, I don't know if we should remove them either (since they will be used at some point)

jkwak-work commented on tests/glsl/shader-subgroup-shuffle-relative.slang:
```diff
@@ -0,0 +1,65 @@
+// TEST:SIMPLE(filecheck=CHECK_CPP):  -allow-glsl -stage compute -entry computeMain -target cpp
+
+//TEST(compute, vulkan):COMPARE_COMPUTE(filecheck-buffer=BUF):-vk -compute -entry computeMain -allow-glsl
+//TEST(compute, vulkan):COMPARE_COMPUTE(filecheck-buffer=BUF):-vk -compute -entry computeMain -allow-glsl -emit-spirv-directly
```
This test is failing for me.

ArielG-NV replied:
I don't know why, it works for me

jkwak-work replied:
Here is the error message,
```
error: slang-test: tests/glsl/shader-subgroup-shuffle-relative.slang:63:10: error: BUF: expected string not found in input
```

jkwak-work commented on tests/glsl/shader-subgroup-shuffle-relative.slang:
```diff
@@ -0,0 +1,65 @@
+//TEST(compute, vulkan):COMPARE_COMPUTE(filecheck-buffer=BUF):-vk -compute -entry computeMain -allow-glsl
+//TEST(compute, vulkan):COMPARE_COMPUTE(filecheck-buffer=BUF):-vk -compute -entry computeMain -allow-glsl -emit-spirv-directly
+
+//#define TEST_when_doubles_dont_crash
```
If this should be applied only to SPIR-V target, you should wrap this like
```
#if !defined(TARGET_SPIRV)
#define TEST_when_doubles_dont_crash
#endif
```

ArielG-NV replied:
it crashes both

jkwak-work replied:
I said it on my previous comments, but repeating here.
The syntax of this test file is in GLSL and it doesn't make sense if the test fails for targeting GLSL.
If GLSL fails, you have to debug and figure out what is going on.

jkwak-work commented on tests/glsl/shader-subgroup-shuffle.slang:
```diff
@@ -0,0 +1,65 @@
+                            subgroupShuffle((double)1, 1) == 1.0 && // crashes, may break spir-v spec
+#endif
+                           subgroupShuffle((float)1, 1) == 1.0f &&
+                           subgroupShuffle((int)1, 1) == 1 &&
```
genIType means not only int but also ivec2, ivec3, or ivec4.
You need to test the vector versions of the function.
```
subgroupShuffle(int(1));
subgroupShuffle(ivec2(1));
subgroupShuffle(ivec3(1));
subgroupShuffle(ivec4(1));
```

ArielG-NV replied:
I did not add any vector types support for subgroup-shuffle, subgroup-shuffle-relative, subgroup-arithmetic, subgroup-shuffle, subgroup_clustered, subgroup_quad; all of which can use

ArielG-NV replied:
and creating any reasonable test for these types is fairly cumbersome at 40 different calls which can take: matrix, vector, scalar, of sint, uint, float, double, half, bool (due to the current generic system)

I really don't understand if I should just be tasked with making this a less cumbersome process rather than implementing 600+ tests to handle these cases

jkwak-work replied:
It sounds like you are not familiar with "Test Driven Development".
When you write tests, you write them as if all of them implementation is already done.
Since the implementation is not done at the time you wrote the tests, most, if not all, tests will fail.
Then you start implement the actual code to make the tests working.

When you said "I did not add any vector type support, (and that's why I am not testing them)" is not a correct order of the approach.

jkwak-work replied:
You are not expected to test those functions against all of available types.
You are expected to test what the spec document requires.

To be more specific, here is an example of "subgroupSuffle".
The doc says these function signatures are required.
```
        genType  subgroupShuffle(genType value,  uint id);
        genIType subgroupShuffle(genIType value, uint id);
        genUType subgroupShuffle(genUType value, uint id);
        genBType subgroupShuffle(genBType value, uint id);
        genDType subgroupShuffle(genDType value, uint id);
```
None of these take matrix and half types.
```
"genType" means float, vector<float, 2>, vector < float, 3 >, vector < float, 4>
"genIType" means int, vector<int, 2>, vector < int, 3 >, vector < int, 4>
"genUType" means uint, vector<uint, 2>, vector < uint, 3 >, vector < uint, 4>
"genBType" means bool, vector< bool, 2>, vector < bool, 3 >, vector < bool, 4>
"genDType" means double, vector< double, 2>, vector < double, 3 >, vector < double, 4>

```
To make the test less redundant, you have to use __generic.
```
__generic<T:__BuiltinType, let N:int>
bool testSubgroupXXX()
{
     typealias genType = vector<T,N>;
     return true
        && subgroupShuffle(genType(0))
        && subgroupShuffleXor(genType(0), 0)
        ....
         ;
}

void computeMain(...)
{
    bool result = true
       && testSubgroupXXX < float, 2 >
        && testSubgroupXXX < float, 3>
        && testSubgroupXXX < float, 4>
        && testSubgroupXXX < uint, 2>
        && testSubgroupXXX < uint, 3>
        && testSubgroupXXX < uint, 4>
        ...
#if defined(Test_when_double_type_works)
        && testSubgroupXXX < double, 3>
        && testSubgroupXXX < double, 3>
        && testSubgroupXXX < double, 4>
#endif // #if defined(Test_when_double_type_works)
        ;
```
This example doesn't show how to test scalar types and you need to include them.
Please check another test file as a more detailed example.
https://github.com/shader-slang/slang/blob/master/tests/glsl-intrinsic/intrinsic-basic.slang#L257

Please make the same changes to the rest of test files too.

Also note that you should have written this test BEFORE you started implementing your code in glsl.meta.slang.

jkwak-work reviewed (commented):
I left more comments.

csyonghe committed:
Support link-time constants. (#3564)

* Support link-time constants.

* Fix.

* Fix.

jkwak-work commented on tests/glsl/shader-subgroup-arithmetic.slang:
```diff
@@ -0,0 +1,99 @@
+    //BUF-NEXT: 1
+    //BUF-NEXT: 1
+    //BUF-NEXT: 1
+    //BUF-NEXT: 1
```
I think these are too verbose and doesn't seem useful.
Let's make it one output that combine all of them with logicalAnd.
```
bool result = true
    && subgroupXXX == YY)
    && (subgroupZZZ == WW)
    && ...
    ;
outputBuffer.data[0] = float(result);
```
Please make the same change to the rest of test files.

jkwak-work commented on tests/glsl/shader-subgroup-arithmetic.slang:
```diff
@@ -0,0 +1,99 @@
+//TEST:SIMPLE(filecheck=CHECK_SPV):  -allow-glsl -stage compute -entry computeMain -target spirv
+
+// not testing hlsl due to missing impl
+// TEST:SIMPLE(filecheck=CHECK_HLSL): -allow-glsl -stage compute -entry computeMain -target hlsl
```
When targeting HLSL, or others if possible, can you narrow down more to which tests are causing problems?
I don't think ALL of arithmetic functions are having problems for HLSL.
If there are some functions working for HLSL, we should test them instead of disabling the entire HLSL test.

jkwak-work reviewed (commented):
I left more comments.

jkwak-work commented:
Can you get the Top Of the Tree from shader-slang/master and rebase to your repository?
Your current change has a merge conflict which prevent CI to run.

jkwak-work reviewed (commented):
I left the error messages for the failing tests.

csyonghe committed:
Fix spirv legalization of nested ararys. (#3567)

* Fix spirv legalization of nested ararys.

* Fix test.

csyonghe committed:
Fix type checking around generic array types. (#3568)

ArielG-NV committed:
redesign tests under suggestions that they should be smaller, more maintainable, and test the most amount of data reasonabley possible (balance with fast iterations);

optional double testing

varying parameter testing

most tests chain results now

ArielG-NV committed:
Merge branch 'glsl.meta.slang_khr_shader_subgroup' of https://github.com/ArielG-NV/slang into glsl.meta.slang_khr_shader_subgroup

ArielG-NV committed:
fix missing impl and merge conflict resolutions

ArielG-NV committed:
Merge branch 'master' of https://github.com/ArielG-NV/slang into glsl.meta.slang_khr_shader_subgroup

ArielG-NV committed:
reundant test code cleanup and organization

move tests to proper location (glsl-intrinsic)

clean up redundant code (input buffers)

ArielG-NV committed:
Merge branch 'shader-slang:master' into glsl.meta.slang_khr_shader_subgroup

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/glsl-test-utils.slang:
```diff
@@ -0,0 +1,56 @@
+//why macros and not templated generics? -- macros are faster to iterate (development) on
```
Please avoid using macros on unit tests. One of the most important aspect of the unit test is that it must be easy to read. The use of macro makes things much harder for the other people to read and maintain.

ArielG-NV replied:
does slang support some version of function pointer/capture passing?

ArielG-NV replied:
(want to have a method to follow the "iterate fast but safety [test oriented] first" model)

jkwak-work replied:
There might be a way to use pointer but you don't need it here.
It will be unnecessary complexity.

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/shader-subgroup-clustered.slang:
```diff
@@ -0,0 +1,47 @@
+                           testFunctionIn2ParamPlus(subgroupClusteredMax, 1, v, 1) &&
+                           testFunctionInInt2ParamPlus(subgroupClusteredAnd, 1, v, 1) &&
+                           testFunctionInInt2ParamPlus(subgroupClusteredOr, 1, v, 1) &&
+                           testFunctionInInt2ParamPlus(subgroupClusteredXor, 1, v, 1);
```
In a unit test code, we should be able to individually enable and disable each and every test as we progressively implement functions. Your way to wrapping a group of tests in to a macro is not preferable.
The use of "if" and "for" are also not preferable in unit test. If possible avoid using any complex logics and unroll them as much as possible so that things are easy to read, and harder to make a mistake. If unit tests start having complexity, we will have to have another unit tests that test the unit tests, which becomes unmaintainable.

jkwak-work commented on tests/glsl-intrinsic/shader-subgroup/shader-subgroup-arithmetic.slang:
```diff
@@ -0,0 +1,131 @@
+    vec[10] = testFunctionIn1Param(subgroupExclusiveMax, 1, test);
+    vec[11] = testFunctionInInt1Param(subgroupExclusiveAnd, 1, test2);
+    vec[12] = testFunctionInInt1Param(subgroupExclusiveOr, 1, test2);
+    vec[13] = testFunctionInInt1Param(subgroupExclusiveXor, 1, test2);
```
The use of the variable "test" and "test2" is unnecessary here. It makes the reader to scroll up and down, which indicates a room for a mistake and unnecessary complexity. Just use a hard coded value to make it easier to read.
If you are not familiar with writing unit tests, my advice may seem counter productive because it is opposite of the coding style of the regular codes. But the purpose of unit tests is different from the regular programing.

ArielG-NV committed:
add missing logical operands support (and remove hlsl/cuda code reuse due to the functional differences) under all And, Or, Xor ops

redesign tests to conform to a better testing paradigm

jkwak-work commented:
Closing this PR and reopening a new one.
